[inductor] make conv lowering work with dynamic shapes

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126823

Fix an issue reported by internal user that conv lowering does not work well with dynamic shapes.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

shunting314 committed:
[inductor] make conv lowering work with unbacked symint

[ghstack-poisoned]

ezyang commented on torch/_inductor/kernel/conv.py:
```diff
@@ -363,6 +365,13 @@ def channels_last_conv():
         and groups == 1
+        # We tried free_unbacked_symbols first but inductor somehow get
+        # unbacked symbols having 's' as prefix and those are not
+        # counted as unbacked symbols by free_unbacked_sybmols function.
```
remove this comment

ezyang commented on torch/_inductor/kernel/conv.py:
```diff
@@ -363,6 +365,13 @@ def channels_last_conv():
+        # counted as unbacked symbols by free_unbacked_sybmols function.
+        #
+        # Call free_symbols instead. The downside is for some cases, we
+        # conservatively not converting conv1x1 to mm. Not a big deal.
```
Because these are backed symints, you have another option, which is to (1) check the hint and (2) add a guard. Also, in both backed and unbacked case, you can check if it's statically known that the numel is > 0. In particular, because of 0/1 specialization, you will nearly always know that the dynamic shapes in question are > 0 and therefore the optimization should apply. So this is basically free (no extra guards) and I would greatly prefer to preserve this optimization even with dynamic shapes.

shunting314 replied:
I can keep this optimization for dynamic shapes. But there is still one thing puzzles me.

From the symbol prefix ('s' here), we know they are backed symint. But those symbols are the number of non-zeros from some integer tensors in my unit test. They are data dependent and seems more like unbacked symint. Any idea what's going on here?

jansel reviewed (approved):
lgtm but fix @ezyang's comments

shunting314 committed:
Update on "[inductor] make conv lowering work with dynamic shapes"


Fix an issue reported by internal user that conv lowering does not work well with dynamic shapes.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 commented on torch/_inductor/kernel/conv.py:
```diff
@@ -363,6 +365,13 @@ def channels_last_conv():
+        #
+        # Call free_symbols instead. The downside is for some cases, we
+        # conservatively not converting conv1x1 to mm. Not a big deal.
+        and not free_symbols(x.get_size())
```
I changed to call `statically_known_gt` iinstead

shunting314 commented:
@pytorchbot merge
